server/services/v1/query_runner.go

```diff
@@ -150,62 +153,129 @@ func (runner *BaseQueryRunner) GetCollections(db *metadata.Database, collName st
+		} else {
+			// double quotes only needed on key, for auto-generated strings we are already adding it in generator
+			keyValueResp = []byte(fmt.Sprintf(`"%s":%s`, v.FieldName, jsonVal))
+		}
+		if len(keysReturnedInResp) > 0 {
+			keysReturnedInResp = append(keysReturnedInResp, []byte(`,`)...)
+			keysReturnedInResp = append(keysReturnedInResp, keyValueResp...)
+		} else {
+			keysReturnedInResp = keyValueResp
+		}
 	}
 	if len(appendTo) == 0 {
-		return nil, ulog.CE("missing index key column(s)")
+		return nil, nil, nil, false, ulog.CE("missing index key column(s)")
 	}
-	return appendTo, nil
+
+	var shouldUseInsert = false
+	if len(autoGeneratedTypes) > 0 {
+		// if we have autogenerated pkey and if it is prone to conflict then force to use Insert API
+		if len(autoGeneratedTypes) == 1 &&
+			autoGeneratedTypes[0] == schema.Int64Type ||
+			autoGeneratedTypes[0] == schema.DateTimeType {
+			shouldUseInsert = true
+		}
+	}
+	// make it JSON
+	keysReturnedInResp = []byte(fmt.Sprintf(`{%s}`, keysReturnedInResp))
+	return doc, appendTo, keysReturnedInResp, shouldUseInsert, nil
 }
 
-func (runner *BaseQueryRunner) insertOrReplace(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant, db *metadata.Database, coll *schema.DefaultCollection, documents [][]byte, insert bool) (*internal.Timestamp, error) {
+func (runner *BaseQueryRunner) insertOrReplace(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant, db *metadata.Database, coll *schema.DefaultCollection, documents [][]byte, insert bool) (*internal.Timestamp, [][]byte, error) {
 	var err error
 	var ts = internal.NewTimestamp()
+	var allKeys [][]byte
 	for _, doc := range documents {
 		var deserializedDoc map[string]interface{}
 		dec := jsoniter.NewDecoder(bytes.NewReader(doc))
 		dec.UseNumber()
 		if err = dec.Decode(&deserializedDoc); ulog.E(err) {
-			return nil, err
+			return nil, nil, err
 		}
 		if err = coll.Validate(deserializedDoc); err != nil {
 			// schema validation failed
-			return nil, err
+			return nil, nil, err
 		}
 
-		indexParts, err := runner.extractIndexParts(coll.Indexes.PrimaryKey.Fields, doc)
-		if ulog.E(err) {
-			return nil, err
+		encodedTable, err := runner.encoder.EncodeTableName(tenant.GetNamespace(), db, coll)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		var indexParts []interface{}
+		var keyReturnedInResp []byte
+		var shouldUseInsert bool
+		if doc, indexParts, keyReturnedInResp, shouldUseInsert, err = runner.extractIndexParts(ctx, encodedTable, coll.Indexes.PrimaryKey.Fields, doc); ulog.E(err) {
+			return nil, nil, err
 		}
 
 		var key keys.Key
-		if key, err = runner.encoder.EncodeKey(tenant.GetNamespace(), db, coll, coll.Indexes.PrimaryKey, indexParts); ulog.E(err) {
-			return nil, err
+		if key, err = runner.encoder.EncodeKey(encodedTable, coll.Indexes.PrimaryKey, indexParts); ulog.E(err) {
+			return nil, nil, err
 		}
 
 		tableData := internal.NewTableDataWithTS(ts, nil, doc)
+		if shouldUseInsert {
```

Comment:
Why not use this variable directly in below if?

Reply:
We are only changing the method call if it is autoGenerate and field is int64/timestamp. Otherwise we need to honor what is passed in the API call. For example, API caller can pass insert true but `shouldUseInsert ` is false because the autoGenerate is not for this field. I can rename this variable to make it more clear.

Reply:
got it.
Then just use it like `if insert || shouldUseInsert` {` in the next if.

Reply:
yeah, right!